README.md
# ua-parser-php #

`ua-parser-php` is a PHP-based pseudo-port of the [ua-parser](http://code.google.com/p/ua-parser/) project. `ua-parser-php`
utilizes the user agents regex YAML file from ua-parser but otherwise creates its own set of attributes to describe a browser, OS, and device. `ua-parser-php`
was created as a new browser-detection library for the browser- and feature-detection library [Detector](https://github.com/dmolsen/Detector).

If you want, you can [test ua-parser-php](http://uaparser.dmolsen.com/) with your browser.

## Usage ##

Straightforward:

    <?php

       require("UAParser.php");
       $result = UA::parse();

       print $result->full;
       // -> Chrome 16.0.912/Mac OS X 10.6.8

       print $result->browserFull;
       // -> "Chrome 16.0.912"
		
       print $result->browser;
       // -> "Chrome"
		
       print $result->version;
       // -> "16.0.912"
		
       print $result->major;
       // -> 16 (minor, build, & revision also available)
		
       print $result->osFull;
       // -> "Mac OS X 10.6.8"
		
       print $result->os;
       // -> "Mac OS X"
		
       print $result->osVersion;
       // -> "10.6.8"
		
       print $result->osMajor;
       // -> 10 (osMinor, osBuild, & osRevision also available)

       /* 
        * in select cases the device information will also be captured
        */

       print $result->deviceFull;
       // -> "Palm Pixi 1.0"
       
       print $result->device;
       // -> "Palm Pixi"

       print $result->deviceVersion
       // -> "1.0"

       print $result->deviceMajor;
       // -> 1 (deviceMinor also available)

       /*
        * Some other generic boolean options
        */

	     print $result->isMobile;
       // -> (would return true if the browser met the criteria of a mobile browser based on the user agent information)

	     print $result->isMobileDevice;
       // -> (would return true if the device met the criteria of a mobile device based on the user agent information)

       print $result->isTablet;
       // -> (would return true if the device was a tablet according to the user agent information)

       print $result->isSpider;
       // -> (would return true if the device was a spider according to the user agent information)

       print $result->isComputer;
       // -> (would return true if the device was a computer according to the user agent information)

		   print $result->isUIWebview;
       // -> (would return true if the user agent was from a uiwebview in ios)

    ?>

If you want to grab a copy of the YAML data from ua-parser each night you can use a cron job and point it at the following bit of code:

    <?php

       require("UAParser.php");
       $result = UA::get();

    ?>

**NOTE:** This will overwrite some changes I've made to the `user_agents_regex.yaml` file included with the `ua-parser-php` distribution.


## Credits ##

Thanks to the ua-parser team for making the YAML file available for others to build upon.

CHANGELOG
THIS CHANGELOG IS AN ATTEMPT TO DOCUMENT CHANGES TO THIS PROJECT.
SUPER MINOR REVISION NUMBERS WON'T SHOW UP IN GITHUB BUT IT'LL HELP
ME ORGANIZE WHEN I MADE A SET OF FIXES

UAP-v1.2.1
 - The following were changes to user_agents_regex.yaml:
 ---- ADD: support for android 4 user agents that have a dash in them

UAP-v1.2.0
 - FIX: should be compatible with PHP 5.2
 - FIX: addressed STRICT mode errors
 - FIX: addressed NOTICE for a missing variable
 - FIX: if isTablet is set to true then isMobile is set to false (mobile to me means phone)
 - THX: Thanks to Mike Bond of WVU Libraries for pointing out the 5.2 incompatibility

UAP-v1.1.0
 - FIX: a number of fixes from bryan shelton
 - The following were changes to user_agents_regex.yaml:
 ---- ADD: support for Chrome Mobile

UAP-v1.0.0
 - just because i don't expect to update this anytime soon and ppl report it's working

UAP-v0.3.1
 - FIX: swapped nil for null in parse()
 - FIX: smarter/saner defaults
 - FIX: now using isset() for family_replacement
 - THX: thanks to bryan shelton for these fixes 

UAP-v0.3.0
 - ADD: can now supply a specific UA to be checked
 - ADD: if the UA contains 'tablet' isTablet is marked true
 - ADD: for certain mobile OSs they report a desktop browser. marking them mobile now.
 - FIX: tablet listing should now be better
 - FIX: the list of mobile browsers was updated
 - FIX: made sure that certain checks won't fail as "false" if a version number was a 0
 - FIX: for the device check, if it returns spider as a result it no longer marks it as mobile
 - FIX: added more mobile browsers to that specific check
 - The following were changes to user_agents_regex.yaml:
 ---- ADD: symphony, minimo, teleca, semc, up.browser, bunjaloo, jasmine, & brew browsers supported
 ---- ADD: windows mobile 6.5 os support
 ---- ADD: amoi, asus, bird, dell, docomo, huawei, i-mate, kyocera, lenovo, lg, microsoft kind,
       motorola, philips, samsung, softbank, & sony ericsson device checks
 ---- FIX: mobile firefox, opera mobile & mini, obigo, polaris, nokiabrowser, ie mobile,
       android, & mobile safari browser checks
 ---- FIX: iOS support
 ---- FIX: htc, android, palm/hp, kindle, ios, generic feature phones & spider device checks

UAP-v0.2.0
 - ADD: added isMobile support
 - ADD: added isTablet support
 - ADD: added isComputer support
 - ADD: added isSpider support

UAP-v0.1.0
 - The following were changes to user_agents_regex.yaml:
 ---- expanded support for Symbia & Nokia Devices, 
 ---- cleaned up some slashies, 
 ---- added Mobile Safari as the family replacement for iOS devices, 
 ---- better support for longer HTC device names
 ---- added feature phones to the device check
 ---- added generic smartphones to the device check
 ---- added AvantGo to the ua check
 ---- tweaked a lot of the nokia checks
 ---- added kindle support to the device section
 ---- added a generic catch for android devices.
 ---- added support for blackberry devices
 ---- changed the blackberry family to 'blackberry webkit' when it's a webkit browser

index.php
<?php

require("UAParser.php");

$result = UA::parse();

print "<html><head><meta name=\"viewport\" content=\"width=device-width\"><title>ua-parser-php Example</title></head><body><h1>ua-parser-php Test</h1>";
print "<p><a href='https://github.com/dmolsen/ua-parser-php'>ua-parser-php</a> is a pseudo-port of <a href='http://code.google.com/p/ua-parser/'>ua-parser</a>. Please use this page to test your browser. ";
print "<a href='https://github.com/dmolsen/ua-parser-php'>ua-parser-php</a> was developed to help support my <a href='http://detector.dmolsen.com/'>Detector project</a>.</p>";

if ($result) {
	print "<pre>";
	foreach($result as $key => $value) {
		print $key.": ".$value."<br />";
	}
	print "</pre>";
} else {
	print "Sorry, ua-parser-php was unable to match your user agent which was: ";
	print $_SERVER["HTTP_USER_AGENT"];
}

print "<h2>Problem?</h2>";
print "<p>If you notice any incorrect information email me at dmolsen+uaparser@gmail.com. Please include the <em>uaOriginal</em> field as well as what you think the browser should be classified as in your message.</p>";
print "</body></html>";

?>

UAParser.php
<?php

/*!
 * ua-parser-php v1.2.1
 *
 * Copyright (c) 2011-2012 Dave Olsen, http://dmolsen.com
 * Licensed under the MIT license
 *
 * ua-parser-php is a PHP-based pseudo-port of the ua-parser project. Learn more about the ua-parser project at:
 * 
 * http://code.google.com/p/ua-parser/
 *
 * The user agents data from the ua-parser project is licensed under the Apache license.
 * spyc-0.5, for loading the YAML, is licensed under the MIT license.
 * The initial list of generic feature phones & smartphones came from Mobile Web OSP under the MIT license
 * The initial list of spiders was taken from Yiibu's profile project under the MIT license.
 *
 */

// address 5.2 compatibility
if (!defined('__DIR__')) define('__DIR__', dirname(__FILE__));
if (!function_exists('json_decode') || !function_exists('json_encode')) {
	require_once(__DIR__."/lib/json/jsonwrapper.php");
}

// load spyc as a YAML loader
require_once(__DIR__."/lib/spyc-0.5/spyc.php");

class UA {
	
	private static $ua;
	private static $accept;
	private static $regexes;
	
	private static $debug = false;
	
	/**
	* Sets up some standard variables as well as starts the user agent parsing process
	*
	* @return {Object}       the result of the user agent parsing
	*/
	public static function parse($ua = NULL) {
		
		self::$ua      = $ua ? $ua : $_SERVER["HTTP_USER_AGENT"];
		self::$accept  = $_SERVER["HTTP_ACCEPT"];
		self::$regexes = Spyc::YAMLLoad(__DIR__."/resources/user_agents_regex.yaml");
		
		// run the regexes to match things up
		$uaRegexes = self::$regexes['user_agent_parsers'];
		foreach ($uaRegexes as $uaRegex) {
			if ($result = self::uaParser($uaRegex)) {
				$result->uaOriginal = self::$ua;
				break;
			}
		}
		
		// if no browser was found check to see if it can be matched at least against a device (e.g. spider, generic feature phone or generic smartphone)
		if (!$result) {
			if (($result = self::deviceParser()) && ($result->device != 'Spider')) {
				$result->isMobile       = true;
				$result->isMobileDevice = true;	
				$result->uaOriginal     = self::$ua;
			} else if ($result->device == "Spider") {
				$result->isMobile       = false;
				$result->isSpider       = true;
				$result->uaOriginal     = self::$ua;
			}
		}
		
		// still false?! see if it's a really dumb feature phone
		if (!$result) {
			if ((strpos(self::$accept,'text/vnd.wap.wml') > 0) || (strpos(self::$accept,'application/vnd.wap.xhtml+xml') > 0) || isset($_SERVER['HTTP_X_WAP_PROFILE']) || isset($_SERVER['HTTP_PROFILE'])) {
				$result = new stdClass();
				$result->device         = "Generic Feature Phone";
				$result->deviceFull     = "Generic Feature Phone";
				$result->isMobile       = true;
				$result->isMobileDevice = true;
				$result->uaOriginal     = self::$ua;
			}
		}

		// log the results when testing
		if (self::$debug) {
			self::log($result);
		}
		
		return $result;
	}
	
	/**
	* Attemps to see if the user agent matches the regex for this test. If so it populates an obj
	* with properties based on the user agent. Will also try to fetch OS & device properties
	*
	* @param  {Array}        the regex to be tested as well as any extra variables that need to be swapped
	*
	* @return {Object}       the result of the user agent parsing
	*/
	private static function uaParser($regex) {
		
		// tests the supplied regex against the user agent
		if (preg_match("/".str_replace("/","\/",$regex['regex'])."/",self::$ua,$matches)) {

			// Define safe parser defaults
			$defaults = array(
				'isMobileDevice' => false,
				'isMobile' => false,
				'isSpider' => false,
				'isTablet' => false,
				'isComputer' => true,
			);
			// build the obj that will be returned starting with defaults
			$obj = (object) $defaults;

			// build the version numbers for the browser
			$obj->major  = isset($regex['v1_replacement']) ? $regex['v1_replacement'] : $matches[2];
			if (isset($matches[3]) || isset($regex['v2_replacement'])) {
				$obj->minor = isset($regex['v2_replacement']) ? $regex['v2_replacement'] : $matches[3];
			}
			if (isset($matches[4])) {
				$obj->build = $matches[4];
			}
			if (isset($matches[5])) {
				$obj->revision = $matches[5];
			}
			
			// pull out the browser family. replace the version number if necessary
			$obj->browser = isset($regex['family_replacement']) ? str_replace("$1",$obj->major,$regex['family_replacement']) : $matches[1];
			
			// set-up a clean version number
			$obj->version = isset($obj->major) ? $obj->major : "";
			$obj->version = isset($obj->minor) ? $obj->version.'.'.$obj->minor : $obj->version;
			$obj->version = isset($obj->build) ? $obj->version.'.'.$obj->build : $obj->version;
			$obj->version = isset($obj->revision) ? $obj->version.'.'.$obj->revision : $obj->version;
			
			// prettify
			$obj->browserFull = $obj->browser;
			if ($obj->version != '') {
				$obj->browserFull .= " ".$obj->version;
			}
			
			// detect if this is a uiwebview call on iOS
			$obj->isUIWebview = (($obj->browser == 'Mobile Safari') && !strstr(self::$ua,'Safari')) ? true : false;
			
			// check to see if this is a mobile browser
			$mobileBrowsers = array("Firefox Mobile","Opera Mobile","Opera Mini","Mobile Safari","webOS","IE Mobile","Playstation Portable",
			                        "Nokia","Blackberry","Palm","Silk","Android","Maemo","Obigo","Netfront","AvantGo","Teleca","SEMC-Browser",
			                        "Bolt","Iris","UP.Browser","Symphony","Minimo","Bunjaloo","Jasmine","Dolfin","Polaris","BREW","Chrome Mobile");
			foreach($mobileBrowsers as $mobileBrowser) {
				if (stristr($obj->browser, $mobileBrowser)) {
					$obj->isMobile = true;
					break;
				}
			}
					
			// figure out the OS for the browser, if possible
			if ($osObj = self::osParser()) {
				$obj = (object) array_merge((array) $obj, (array) $osObj);
			}
			
			// create an attribute combinining browser and os
			if (isset($obj->osFull) && $obj->osFull) {
				$obj->full = $obj->browserFull."/".$obj->osFull;
			}
			
			// figure out the device name for the browser, if possible
			if ($deviceObj = self::deviceParser()) {
				$obj = (object) array_merge((array) $obj, (array) $deviceObj);
			}
			
			// if this is a mobile browser make sure mobile device is set to true
			if ($obj->isMobile) {
				$obj->isMobileDevice = true; // this is going to catch android devices
			} else if ($obj->isMobileDevice) {
				$obj->isMobile = true;       // this will catch some weird htc devices
			}
			
			// if OS is Android check to see if this is a tablet. won't work on UA strings less than Android 3.0
			if ((isset($obj->os) && $obj->os == 'Android') && !strstr(self::$ua, 'Mobile')) {
				$obj->isTablet = true;
				$obj->isMobile = false;
			}
			
			// some select mobile OSs report a desktop browser. make sure we note they're mobile
			$mobileOSs = array('Windows Phone 6.5','Windows CE','Symbian OS');
			if (isset($obj->os) && in_array($obj->os,$mobileOSs)) {
				$obj->isMobile       = true;
				$obj->isMobileDevice = true;
			}
			
			if (stristr(self::$ua,"tablet")) {
				$obj->isTablet       = true;
				$obj->isMobileDevice = true;
				$obj->isMobile       = false;
			}
			
			// record if this is a spider
			$obj->isSpider = (isset($obj->device) && $obj->device == "Spider") ? true : false;
			
			// record if this is a computer
			$obj->isComputer = (!$obj->isMobile && !$obj->isSpider && !$obj->isMobileDevice) ? true : false;
				
			return $obj;
			
		} else {
			return false;
		}
	}
	
	/**
	* If the user agent is matched in uaParser() it also tries to check the OS and get properties
	*
	* @return {Object}       the result of the os parsing
	*/
	private static function osParser() {
		
		// build the obj that will be returned
		$osObj = new stdClass();
		
		// run the regexes to match things up
		$osRegexes = self::$regexes['os_parsers'];
		foreach ($osRegexes as $osRegex) {
			if (preg_match("/".str_replace("/","\/",$osRegex['regex'])."/",self::$ua,$matches)) {
				
				// Make sure matches 2 and 3 are at least set to null for setting
				// Major and Minor defaults
				if (!isset($matches[2])) { $matches[2] = null; }
				if (!isset($matches[3])) { $matches[3] = null; }

				// basic properties
				$osObj->osMajor   = isset($osRegex['os_v1_replacement']) ? $osRegex['os_v1_replacement'] : $matches[2];
				$osObj->osMinor   = isset($osRegex['os_v2_replacement']) ? $osRegex['os_v2_replacement'] : $matches[3];
				if (isset($matches[4])) {
					$osObj->osBuild = $matches[4];
				}
				if (isset($matches[5])) {
					$osObj->osRevision = $matches[5];
				}
				$osObj->os        = isset($osRegex['os_replacement'])    ? str_replace("$1",$osObj->osMajor,$osRegex['os_replacement'])  : $matches[1];

				// os version
				$osObj->osVersion = isset($osObj->osMajor) ? $osObj->osMajor : "";
				$osObj->osVersion = isset($osObj->osMinor) ? $osObj->osVersion.'.'.$osObj->osMinor : $osObj->osVersion;
				$osObj->osVersion = isset($osObj->osBuild) ? $osObj->osVersion.'.'.$osObj->osBuild : $osObj->osVersion;
				$osObj->osVersion = isset($osObj->osRevision) ? $osObj->osVersion.'.'.$osObj->osRevision : $osObj->osVersion; 
				
				// prettify
				$osObj->osFull    = $osObj->os." ".$osObj->osVersion;
				
				return $osObj;
			}
		}
		return false;
	}
	
	/**
	* If the user agent is matched in uaParser() it also tries to check the device and get its properties
	*
	* @return {Object}       the result of the device parsing
	*/
	private static function deviceParser() {
		
		// build the obj that will be returned
		$deviceObj = new stdClass();
		
		// run the regexes to match things up
		$deviceRegexes = self::$regexes['device_parsers'];
		foreach ($deviceRegexes as $deviceRegex) {
			if (preg_match("/".str_replace("/","\/",$deviceRegex['regex'])."/i",self::$ua,$matches)) {
				
				// Make sure matches 2 and 3 are at least set to null for setting
				// Major and Minor defaults
				if (!isset($matches[2])) { $matches[2] = null; }
				if (!isset($matches[3])) { $matches[3] = null; }

				// basic properties
				$deviceObj->deviceMajor  = isset($deviceRegex['device_v1_replacement']) ? $deviceRegex['device_v1_replacement'] : $matches[2];
				$deviceObj->deviceMinor  = isset($deviceRegex['device_v2_replacement']) ? $deviceRegex['device_v2_replacement'] : $matches[3];
				$deviceObj->device       = isset($deviceRegex['device_replacement']) ? str_replace("$1",$matches[1],$deviceRegex['device_replacement']) : str_replace("_"," ",$matches[1]);
				
				// device version?
				$deviceObj->deviceVersion = isset($deviceObj->deviceMajor) ? $deviceObj->deviceMajor : "";
				$deviceObj->deviceVersion = isset($deviceObj->deviceMinor) ? $deviceObj->deviceVersion.'.'.$deviceObj->deviceMinor : $deviceObj->deviceVersion;
				
				// prettify
				$deviceObj->deviceFull = $deviceObj->device." ".$deviceObj->deviceVersion;
				
				// check to see if this is a mobile device
				// this isn't really needed because if it matches a mobile browser it'll automatically mark it as a mobile device
				$deviceObj->isMobileDevice = false;
				$mobileDevices  = array("iPhone","iPod","iPad","HTC","Kindle","Lumia","Amoi","Asus","Bird","Dell","DoCoMo","Huawei","i-mate","Kyocera",
				                        "Lenovo","LG","Kin","Motorola","Philips","Samsung","Softbank","Palm","HP ","Generic Feature Phone","Generic Smartphone");
				foreach($mobileDevices as $mobileDevice) {
					if (stristr($deviceObj->device, $mobileDevice)) {
						$deviceObj->isMobileDevice = true;
						break;
					}
				}

				// check to see if this is a tablet (not perfect)
				$deviceObj->isTablet = false;
				$tablets = array("Kindle","iPad","Playbook","TouchPad","Dell Streak","Galaxy Tab","Xoom");
				foreach($tablets as $tablet) {
					if (stristr($deviceObj->device, $tablet)) {
						$deviceObj->isTablet = true;
						break;
					}
				}
				
				return $deviceObj;
			}
		}
		return false;
	}
	
	/**
	* Logs the user agent info
	*/
	private static function log($data) {
		if (!$data) {
			$data = new stdClass();
			$data->ua = self::$ua;
		} 
		$jsonData = json_encode($data);
		$fp = fopen(__DIR__."/log/user_agents.log", "a");
		fwrite($fp, $jsonData."\r\n");
		fclose($fp);
	}
	
	/**
	* Gets the latest user agent. Back-ups the old version first. it will fail silently if something is wrong...
	*/
	public static function get() {
		if ($data = @file_get_contents("http://ua-parser.googlecode.com/svn/trunk/resources/user_agent_parser.yaml")) {
			copy(__DIR__."/resources/user_agents_regex.yaml", __DIR__."/resources/user_agents_regex.".date("Ymdhis").".yaml");
			$fp = fopen(__DIR__."/resources/user_agents_regex.yaml", "w");
			fwrite($fp, $data);
			fclose($fp);
		}
	}
}